### PR TITLE
AddFeature Modification

### PR DESCRIPTION
**Pull request by s9latimm**

### CODE REVIEW

File: include/vara/Feature/FeatureModel.h [hunk 1]
```diff
@@ -154,28 +153,29 @@ class FeatureModelBuilder : private FeatureModel {
...
+      return nullptr;
+    }
+    return Features[FeatureName].get();
   }
 
-  bool addFeature(Feature &F);
-
-  FeatureModelBuilder *addParent(const std::string &FeatureName,
-                                 const std::string &ParentName) {
+  FeatureModelBuilder *addEdge(const std::string &ParentName,
```
Comment (vulder) [inline include/vara/Feature/FeatureModel.h:177]: Why did you rename this? `addParent` seemed fine to me

File: include/vara/Feature/FeatureModelTransaction.h [hunk 2]
```diff
@@ -135,14 +135,13 @@ class FeatureModelModification {
 
 protected:
   /// \brief Set the parrent of a Feature.
-  static void setParent(Feature &F, Feature *Parent) {
-    F.setParent(Parent);
-  }
+  static void setParent(Feature &F, Feature *Parent) { F.setParent(Parent); }
 
-  static void addChild(Feature &F, Feature &Child) {
+  static void addChild(Feature &F, Feature *Child) {
```
Comment (vulder) [inline include/vara/Feature/FeatureModelTransaction.h]:
```suggestion
  static void addChild(Feature &F, Feature &Child) {
```
There is no go way to handle nullptrs here, so why should `Child` be a ptr?

File: lib/Feature/FeatureModel.cpp [hunk 3]
```diff
@@ -14,28 +14,15 @@ void FeatureModel::dump() const {
...
 
-bool FeatureModel::addFeature(std::unique_ptr<Feature> NewFeature) {
-  // TODO(s9latimm): check consistency
+Feature *FeatureModel::addFeature(std::unique_ptr<Feature> NewFeature) {
   auto FeatureName = std::string(NewFeature->getName());
   if (!Features.try_emplace(FeatureName, std::move(NewFeature)).second) {
-    return false;
+    return nullptr;
   }
   auto *InsertedFeature = Features[FeatureName].get();
```
Comment (vulder) [inline lib/Feature/FeatureModel.cpp]: If I remember correctly `try_emplace` returns and iterator, bool pair.
You create a local var with this and then use it to make the check and you don't have to lookup the feature.
```
auto PosInsertedFeature = Features.try_emplace(...)
if (PosInsertedFeature.second) {
  return nullptr;
}
...
```

File: lib/Feature/FeatureModelParser.cpp [hunk 4]
```diff
@@ -30,7 +30,9 @@ bool FeatureModelXmlParser::parseConfigurationOption(xmlNode *Node,
       } else if (!xmlStrcmp(Head->name, XmlConstants::OPTIONAL)) {
         Opt = Cnt == "True";
       } else if (!xmlStrcmp(Head->name, XmlConstants::PARENT)) {
-        FMB.addParent(Name, Cnt);
+        // TODO(s9latimm) fix me -- breaks if Name not set / parsed yet
```
Comment (vulder) [inline lib/Feature/FeatureModelParser.cpp:33]: What should we do in this case?

File: unittests/Feature/FeatureModel.cpp [hunk 5]
```diff
@@ -48,14 +50,9 @@ TEST(FeatureModel, cloneRoot) {
 
 TEST(FeatureModel, cloneRelationship) {
   FeatureModelBuilder B;
-  BinaryFeature AA("aa");
-  BinaryFeature AB("ab");
-  auto CS = Feature::NodeSetType();
-  CS.insert(&AA);
-  CS.insert(&AB);
-  B.addParent("aa", "root")->addFeature(AA);
-  B.addParent("ab", "root")->addFeature(AB);
```
Comment (vulder) [inline unittests/Feature/FeatureModel.cpp:57]: I guess the parent relation is not strictly needed in this test or?

File: unittests/Feature/FeatureModelTransaction.cpp [hunk 6]
```diff
@@ -16,42 +16,58 @@ class FeatureModelModificationTest : public ::testing::Test,
...
   void SetUp() override {
     FeatureModelBuilder B;
-    B.addParent("a", "root")->makeFeature<BinaryFeature>("a", true);
+    B.makeFeature<BinaryFeature>("a");
     FM = B.buildFeatureModel();
   }
 
   // Dummy method to fullfill the FeatureModelModification interface
-  void exec(FeatureModel &FM) override {};
+  void exec(FeatureModel &FM) override{};
```
Comment (vulder) [inline unittests/Feature/FeatureModelTransaction.cpp:24]:
```suggestion
  void exec(FeatureModel &FM) override {};
```
What version of clang-format are you using?

**Review by vulder — CHANGES_REQUESTED** [review]
Did you adapt the FeatureModelBuilder to also use the modifications? Where there problems?

File: lib/Feature/FeatureModelParser.cpp (hunk 4 above)
Comment (s9latimm) [reply lib/Feature/FeatureModelParser.cpp:33]: We could search `name`  beforehand with an initial pass over all `Node->children`, unfortunately that way we have to look at each node twice (or at least until we find `name`). I will add a separate PR from `vara-dev` with the fix, so that we can resolve this quickly.

File: include/vara/Feature/FeatureModel.h (hunk 1 above)
Comment (s9latimm) [reply include/vara/Feature/FeatureModel.h:177]: I wanted to clarify, that calling this implies, that the children relation between these nodes will be updated too (while building the tree) -- in contrast to the `FeatureModelModification` where we implemented `setParent` so it literally changes only that.

File: unittests/Feature/FeatureModel.cpp (hunk 5 above)
Comment (s9latimm) [reply unittests/Feature/FeatureModel.cpp:57]: As we add those features without changing root, those edges are already implied (so is the existence of a root node, which it neither of those two).

File: unittests/Feature/FeatureModelTransaction.cpp (hunk 6 above)
Comment (s9latimm) [reply unittests/Feature/FeatureModelTransaction.cpp:24]: I updated today to `13.0.0` (latest from `llvm-project/main`) but above change is still persistent (with even more similar diffs according to changed WS rules).

File: include/vara/Feature/FeatureModelTransaction.h (hunk 2 above)
Comment (s9latimm) [reply include/vara/Feature/FeatureModelTransaction.h]: As we talked about it last week already -- there is no reason, as any `nullptr` should be handled beforehand -- will revert that and update references accordingly.

File: unittests/Feature/FeatureModelTransaction.cpp (hunk 6 above)
Comment (vulder) [reply unittests/Feature/FeatureModelTransaction.cpp:24]: hmm, kk
